File: Leetcode/add_two_numbers.py
# https://leetcode.com/problems/add-two-numbers/
'''
    Author - Subhajit Das
    University of Engineering and Management, Kolkata
    30/05/2019
'''


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        digit1 = l1
        digit2 = l2
        carry = 0
        result = None
        start = None

        while digit1 != None and digit2 != None:
            digit3 = digit1.val + digit2.val + carry
            (digit3, carry) = (digit3, 0) if digit3 < 10 else (digit3 % 10, 1)
            node = ListNode(digit3)
            if result == None:
                result = node
                start = result
            else:
                result.next = node
                result = result.next
            digit1 = digit1.next
            digit2 = digit2.next

        if digit1 == None:
            while digit2 != None:
                digit3 = carry + digit2.val
                (digit3, carry) = (digit3, 0) if digit3 < 10 else (digit3 % 10, 1)
                node = ListNode(digit3)
                result.next = node
                result = result.next
                digit2 = digit2.next

        if digit2 == None:
            while digit1 != None:
                digit3 = carry + digit1.val
                (digit3, carry) = (digit3, 0) if digit3 < 10 else (digit3 % 10, 1)
                node = ListNode(digit3)
                result.next = node
                result = result.next
                digit1 = digit1.next

        if carry == 1:
            node = ListNode(carry)
            result.next = node

        return start

File: Leetcode/AddTwoNumbers.java
// https://leetcode.com/problems/add-two-numbers/
/*
    Author @Subhajit Das (sdsubhajitdas.github.io)
    SWE @Turbot HQ India PVT Ltd.
    25/07/2021
*/

import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return this.addTwoNumbersHelper(l1, l2, 0);
    }

    private ListNode addTwoNumbersHelper(ListNode firstDigit, ListNode secondDigit, int carryOver) {

        if(firstDigit == null && secondDigit == null) {
            return carryOver > 0 ? new ListNode(carryOver) : null;
        }


        int firstDigitValue = firstDigit != null ? firstDigit.val : 0;
        int secondDigitValue = secondDigit != null ? secondDigit.val : 0;

        ListNode resultDigit = new ListNode(firstDigitValue+secondDigitValue+carryOver);

        resultDigit.next = this.addTwoNumbersHelper(
            firstDigit!=null ? firstDigit.next : null,
            secondDigit!=null ? secondDigit.next : null,
            resultDigit.val/10
        );
        resultDigit.val = resultDigit.val%10;

        return resultDigit;
    }
}